### PR TITLE
Change backup driver to clean up snapshot/clonefromsnapshot/deletesnapshot cr which reaches terminal status

### PR DESCRIPTION
**Pull request by xinyanw409**
Remove snapshot/clonefromsnapshot/deletesnapshot when its state reaches:
-SnapshotPhaseUploaded
-ClonePhaseCompleted
-DeleteSnapshotPhaseCompleted
for more than 24 hours.

https://container-dp.svc.eng.vmware.com/job/Container_Precheck_Velero/587/

Manually test and check the log to track snapshot/clonefromsnapshot/deletesnapshot completion time and deletion status

```
time="2020-10-21T10:56:32Z" level=info msg="DeleteSnapshot CR delete-7cb06bbf-6b69-4045-8a34-325d6f950b70 reaches phase Completed more than 4 hours, deleting this CR." controller=BackupDriverController logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver/backup_driver_controller_base.go:722"
time="2020-10-21T10:56:32Z" level=info msg="DeleteSnapshot CR delete-affac91a-46da-434a-aa0a-c2c96530ec57 reaches phase Completed more than 4 hours, deleting this CR." controller=BackupDriverController logSource="/go/src/github.com/vmware-tanzu/velero-plugin-for-vsphere/pkg/backupdriver/backup_driver_controller_base.go:722"
```

### CODE REVIEW

**Review by swgupta-vmw — COMMENTED** [review]
In case of guest cluster, how do we clean up the corresponding CRs in the supervisor cluster? Instead of being cleaned up by the supervisor backup driver, should they be cleaned up at the same time the guest CRs are cleaned?

File: pkg/backupdriver/backup_driver_controller_base.go [hunk 1]
```diff
@@ -396,6 +396,17 @@ func (ctrl *backupDriverController) syncSnapshotByKey(key string) error {
 
 	if snapshot.Status.Phase != backupdriverapi.SnapshotPhaseNew {
 		ctrl.logger.Debugf("Skipping snapshot, %v, which is not in New phase. Current phase: %v", key, snapshot.Status.Phase)
+		if snapshot.Status.Phase == backupdriverapi.SnapshotPhaseUploaded {
+			// If Snapshot CR status reaches terminal state more than 1 hour, Snapshot CR should be deleted
+			now := time.Now()
+			if snapshot.Status.CompletionTimestamp.Add(time.Hour).Unix() < now.Unix() {
```
Comment (xing-yang) [inline pkg/backupdriver/backup_driver_controller_base.go:402]: I don't think this 1 hour should be hard-coded.  It should be configurable.  Also default of 1 hour is too short.

File: pkg/backupdriver/backup_driver_controller_base.go [hunk 2]
```diff
@@ -493,12 +504,22 @@ func (ctrl *backupDriverController) syncCloneFromSnapshotByKey(key string) error
...
-	case backupdriverapi.ClonePhaseCompleted:
 	case backupdriverapi.ClonePhaseFailed:
 	case backupdriverapi.ClonePhaseCanceling:
 	case backupdriverapi.ClonePhaseCanceled:
 		ctrl.logger.Infof("Skipping cloneFromSnapshot %s which is not in New or Retry phase. Current phase: %v", key, cloneFromSnapshot.Status.Phase)
 		return nil
+	case backupdriverapi.ClonePhaseCompleted:
+		// If Snapshot CR status reaches terminal state more than 1 hour, Snapshot CR should be deleted
+		now := time.Now()
+		if cloneFromSnapshot.Status.CompletionTimestamp.Add(time.Hour).Unix() < now.Unix() {
```
Comment (xing-yang) [inline pkg/backupdriver/backup_driver_controller_base.go:515]: Same here.  The time period should be configurable, not hard-coded.

File: pkg/backupdriver/backup_driver_controller_base.go [hunk 3]
```diff
@@ -694,6 +715,17 @@ func (ctrl *backupDriverController) syncDeleteSnapshotByKey(key string) error {
 	if delSnapshot.Status.Phase != backupdriverapi.DeleteSnapshotPhaseNew {
 		ctrl.logger.Infof("Skipping DeleteSnapshot, %v, which is not in New phase. Current phase: %v",
 			key, delSnapshot.Status.Phase)
+		if delSnapshot.Status.Phase == backupdriverapi.DeleteSnapshotPhaseCompleted {
+			// If DeleteSnapshot CR status reaches terminal state more than 1 hour, DeleteSnapshot CR should be deleted
+			now := time.Now()
+			if delSnapshot.Status.CompletionTimestamp.Add(time.Hour).Unix() < now.Unix() {
```
Comment (xing-yang) [inline pkg/backupdriver/backup_driver_controller_base.go:721]: Same comment here.

**Comment by xing-yang** [comment]
I thought we discussed about this in the standup meetings that the time period should be configurable, not hard-coded to 1 hour. 

Also please rebase.

**Comment by xinyanw409** [comment]
> In case of guest cluster, how do we clean up the corresponding CRs in the supervisor cluster? Instead of being cleaned up by the supervisor backup driver, should they be cleaned up at the same time the guest CRs are cleaned?

As we discussed in standup meeting, currently we clean up CRs in supervisor cluster and guest cluster separately.

File: pkg/backupdriver/backup_driver_controller_base.go (hunk 1 above)
Comment (xinyanw409) [reply pkg/backupdriver/backup_driver_controller_base.go]: Change to use configurable time period. This change uses the same time constant as https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/186, will merge PR 186 first and resolve the conflicts.

File: pkg/backupdriver/backup_driver_controller_base.go (hunk 2 above)
Comment (xinyanw409) [reply pkg/backupdriver/backup_driver_controller_base.go]: Same as above.

File: pkg/backupdriver/backup_driver_controller_base.go (hunk 3 above)
Comment (xinyanw409) [reply pkg/backupdriver/backup_driver_controller_base.go]: Same as above.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/268/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/272/

File: pkg/backupdriver/backup_driver_controller_base.go [hunk 4]
```diff
@@ -395,6 +396,17 @@ func (ctrl *backupDriverController) syncSnapshotByKey(key string) error {
 
 	if snapshot.Status.Phase != backupdriverapi.SnapshotPhaseNew {
 		ctrl.logger.Debugf("Skipping snapshot, %v, which is not in New phase. Current phase: %v", key, snapshot.Status.Phase)
+		if snapshot.Status.Phase == backupdriverapi.SnapshotPhaseUploaded {
+			// If Snapshot CR status reaches terminal state more than 1 hour, Snapshot CR should be deleted
+			now := time.Now()
+			if now.After(snapshot.Status.CompletionTimestamp.Add(constants.CRCleanUpMaximum * time.Hour)) {
```
Comment (xing-yang) [inline pkg/backupdriver/backup_driver_controller_base.go:402]: Please make changes here accordingly.  I added comments to the other PR.

File: pkg/constants/constants.go [hunk 5]
```diff
@@ -275,3 +275,6 @@ const VMwareSystemVMUUID = "vmware-system-vm-uuid"
 
 // Label to set for Velero to ignore resources
 const VeleroExcludeLabel = "velero.io/exclude-from-backup"
+
+// Maximum time window to clean up CR which is in terminate state
```
Comment (xing-yang) [inline pkg/constants/constants.go:279]: terminate -> terminal

**Review by xing-yang — COMMENTED** [review]
So this is not really configurable.  Looks like more changes are needed if we make this configurable.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/279/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/280/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/283/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/284/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/287/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/288/

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/289/

**Comment by xing-yang** [comment]
LGTM

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/291/

**Review by deepakkinni — APPROVED** [review]
lgtm

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/292/

**Pull request merged**